File: src/Liquidcast.Api/Services/ScheduleMath.cs
using Liquidcast.Api.Models;

namespace Liquidcast.Api.Services;

/// <summary>Shared timing math for scheduled tracks (mirrored client-side in the SPA).</summary>
public static class ScheduleMath
{
    /// <summary>Audible duration of an entry: cue-out (or file end) minus cue-in, never negative.</summary>
    public static double EffectiveDurationSec(ScheduledTrack e)
    {
        var total = e.Track?.DurationSec ?? 0;
        var end = e.CueOutSec is > 0 ? Math.Min(e.CueOutSec.Value, total > 0 ? total : e.CueOutSec.Value) : total;
        var start = e.CueInSec is > 0 ? e.CueInSec.Value : 0;
        return Math.Max(0, end - start);
    }

    public static DateTime EndUtc(ScheduledTrack e) => e.StartUtc.AddSeconds(EffectiveDurationSec(e));

    public static bool IsActive(ScheduledTrack e, DateTime nowUtc) =>
        e.StartUtc <= nowUtc && nowUtc < EndUtc(e);

    /// <summary>Half-open interval overlap — touching entries are adjacent, not overlapping.</summary>
    public static bool Overlaps(DateTime aStart, DateTime aEnd, DateTime bStart, DateTime bEnd) =>
        aStart < bEnd && bStart < aEnd;

    /// <summary>Playback priority: lower is better. Priority 1..4 → 1..4, Fallback line (0) → 5.</summary>
    public static int PriorityKey(ScheduledTrack e) => e.Line == 0 ? 5 : e.Line;
}

File: src/web/src/components/schedule/scheduleMath.js
// Client-side mirror of Services/ScheduleMath.cs — keep the two in sync.

// Audible duration of a (dragged or scheduled) track: cue-out (or file end) minus cue-in.
export function effDurationSec({ durationSec, cueInSec, cueOutSec }) {
  const total = durationSec || 0
  const end = cueOutSec > 0 ? Math.min(cueOutSec, total > 0 ? total : cueOutSec) : total
  const start = cueInSec > 0 ? cueInSec : 0
  return Math.max(0, end - start)
}

// Half-open interval overlap — touching entries are adjacent, not overlapping.
export function overlaps(aStart, aEnd, bStart, bEnd) {
  return aStart < bEnd && bStart < aEnd
}

// Classifies a candidate window against existing entries (excluding its own id).
// - sameLine: overlap on the same line → never allowed.
// - crossLine: any overlap on another line → the override checkbox becomes available.
// - needsOverride: overlap with a plain (non-override) entry on another line → save
//   requires the override flag. Overlapping an override entry is already sanctioned.
export function classifyOverlap(entries, { id, line, startMs, endMs }) {
  let sameLine = false
  let crossLine = false
  let needsOverride = false
  for (const o of entries) {
    if (id != null && o.id === id) continue
    const oS = new Date(o.startUtc).getTime()
    const oE = new Date(o.endUtc).getTime()
    if (!overlaps(startMs, endMs, oS, oE)) continue
    if (o.line === line) sameLine = true
    else { crossLine = true; if (!o.override) needsOverride = true }
  }
  return { sameLine, crossLine, needsOverride }
}

// Display order: fallback line on top, then Priority 1 (highest) … Priority 4.
export const LINE_ORDER = [0, 1, 2, 3, 4]
